File: reference/calc-python/tests/generate_destruction_large_fixtures.py
import json
import random
import os
import sys

# Resolve path to engine
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../engine")))
from destruction_calc_engine import DestructionCalculatorEngine

def generate_large_fixtures():
    engine = DestructionCalculatorEngine()
    
    # Load sp_mapping locally
    sp_mapping_path = os.path.abspath(os.path.join(os.path.dirname(__file__), "../seraphdb_json/skill_sp_mapping.json"))
    sp_mapping = {}
    if os.path.exists(sp_mapping_path):
        with open(sp_mapping_path, "r", encoding="utf-8") as f:
            sp_mapping = json.load(f)
            
    # Get available styles, enemies, skills
    styles = engine.styles
    enemies = engine.enemies
    
    # Filter skills that have attack parts
    skills = []
    for s in engine.skills:
        parts = engine._flatten_parts(s.get("parts", []))
        has_attack = False
        for p in parts:
            if p.get("skill_type") in [
                "AttackNormal", "AttackSkill", "DamageRateChangeAttackSkill",
                "PenetrationCriticalAttack", "PenetrationNormalAttack", "PenetrationSkill",
                "TokenAttack", "AttackBySp", "AttackByOwnDpRate", "FixedHpDamageRateAttack"
            ]:
                has_attack = True
                break
        if has_attack:
            skills.append(s)
            
    # Add special placeholders
    skills.append({"name": "通常攻撃", "id": 46001101})
    skills.append({"name": "追撃", "id": 46001191})
    
    cases = []
    
    for i in range(1000):
        # Choose a random style
        style = random.choice(styles)
        # Choose accessories
        accs = []
        accessory_destruction_rate_bonus = 0.0
        if random.random() < 0.2:
            # Various accessory destruction rate bonuses that exist in-game
            accessory_destruction_rate_bonus = random.choice([0.10, 0.12, 0.15])
            
        # Choose random buffs
        buffs = []
        # destruction buffs
        num_buffs = random.randint(0, 3)
        for _ in range(num_buffs):
            buffs.append({
                "statusType": "DestructionUp",
                "power": round(random.uniform(5.0, 120.0), 2),
                "skillName": f"BuffSkill{random.randint(1, 5)}"
            })
        # ignored buffs
        if random.random() < 0.3:
            buffs.append({
                "statusType": random.choice(["AttackUp", "CritRateUp", "MindEye"]),
                "power": 30.0,
                "skillName": "IgnoredSkill"
            })
            
        # Choose random enemy
        enemy = random.choice(enemies)
        
        # Choose random debuffs
        debuffs = []
        if random.random() < 0.3:
            debuffs.append({
                "statusType": "DefenseDown",
                "power": 30.0,
                "skillName": "DebuffSkill"
            })
            
        # Choose random skill
        skill = random.choice(skills)
        
        # Generate random hits
        num_hits = random.randint(1, 12)
        hits = []
        for _ in range(num_hits):
            hits.append({
                "damage": round(random.uniform(0.0, 150000.0), 2),
                "isMultiHit": random.choice([True, False]),
                "hitRatio": round(random.uniform(0.05, 1.0), 2)
            })
            
        input_data = {
            "attacker": {
                "characterId": style.get("chara", "Unknown"),
                "styleId": style["id"],
                "accessories": [],
                "accessoryDestructionRateBonus": accessory_destruction_rate_bonus,
                "statusEffects": buffs
            },
            "defender": {
                "enemyId": enemy["id"],
                "enemyName": enemy.get("name", "Unknown"),
                "destructionRate": round(random.uniform(1.0, 3.0), 4),
                "destructionLimit": random.choice([None, round(random.uniform(3.0, 5.0), 2)]),
                "dp": round(random.uniform(0.0, 300000.0), 2),
                "destructionResist": round(random.uniform(0.0, 0.5), 4),
                "statusEffects": debuffs
            },
            "skill": {
                "skillId": skill["id"],
                "name": skill["name"]
            },
            "hits": hits
        }
        
        # Inject overrides from sp_mapping
        skill_data = input_data.get("skill", {})
        skill_name = skill_data.get("name")
        
        mapping_info = None
        if skill_name:
            clean_name = skill_name.replace("[単独発動]", "").split("[")[0].split("(")[0].split("（")[0].strip()
            mapping_info = sp_mapping.get(skill_name)
            if not mapping_info:
                mapping_info = sp_mapping.get(clean_name)
                
        if mapping_info:
            sp_val = mapping_info.get("sp")
            if sp_val is not None and sp_val != "-":
                skill_data["spCostOverride"] = float(sp_val)
            else:
                skill_data["spCostOverride"] = 0.0
            skill_data["isNormalAttack"] = mapping_info.get("is_normal_attack", False)
            skill_data["isPursuit"] = mapping_info.get("is_pursuit", False)
        else:
            if skill_name:
                if "通常攻撃" in skill_name:
                    skill_data["isNormalAttack"] = True
                elif "追撃" in skill_name:
                    skill_data["isPursuit"] = True
        input_data["autoBreak"] = True
        
        try:
            expected = engine.calculate_destruction(input_data)
            cases.append({
                "name": f"Random Case {i}",
                "input": input_data,
                "expected": expected
            })
        except Exception as e:
            # Skip cases that crash (e.g. invalid setup/fallback)
            continue
            
    output_path = os.path.join(os.path.dirname(__file__), "fixtures/test_cases_destruction_large.json")
    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(cases, f, indent=2, ensure_ascii=False)
        
    print(f"Generated {len(cases)} randomized destruction test cases in {output_path}")

if __name__ == "__main__":
    generate_large_fixtures()

File: scripts/generate-destruction-large-fixtures.mjs
import { createHash } from 'node:crypto';
import { existsSync, readFileSync, writeFileSync } from 'node:fs';
import path from 'node:path';

import { calculateDestruction, loadDamageCalculationData } from '../src/index.js';

const ROOT_DIR = path.resolve(import.meta.dirname, '..');
const FIXTURE_PATH = path.join(ROOT_DIR, 'tests/fixtures/test_cases_destruction_large.json');
const METADATA_PATH = path.join(ROOT_DIR, 'tests/fixtures/test_cases_destruction_large.meta.json');
const SKILL_MAPPING_PATH = path.join(ROOT_DIR, 'tests/fixtures/skill_sp_mapping.json');
const SYNC_METADATA_PATH = path.join(ROOT_DIR, 'json/_sync_metadata.json');
const SEED = 20260701;
const CASE_COUNT = 1000;
const GENERATOR_VERSION = 1;
const ATTACK_SKILL_TYPES = new Set([
  'AttackNormal',
  'AttackSkill',
  'DamageRateChangeAttackSkill',
  'PenetrationCriticalAttack',
  'PenetrationNormalAttack',
  'PenetrationSkill',
  'TokenAttack',
  'AttackBySp',
  'AttackByOwnDpRate',
  'FixedHpDamageRateAttack',
]);
const CALC_CORE_FILES = [
  'src/domain/damage-calculator.js',
  'src/domain/destruction-calculator.js',
  'src/domain/calculator-helpers.js',
  'src/contracts/damage-calculation.js',
  'src/data/damage-calculation-data.js',
];

function sha256(content) {
  return createHash('sha256').update(content).digest('hex');
}

function hashFile(relativePath) {
  return sha256(readFileSync(path.join(ROOT_DIR, relativePath)));
}

function createPrng(seed) {
  let state = seed >>> 0;
  return () => {
    state += 0x6D2B79F5;
    let value = state;
    value = Math.imul(value ^ (value >>> 15), value | 1);
    value ^= value + Math.imul(value ^ (value >>> 7), value | 61);
    return ((value ^ (value >>> 14)) >>> 0) / 4294967296;
  };
}

function pick(random, values) {
  if (values.length === 0) {
    throw new Error('Cannot choose from an empty collection.');
  }
  return values[Math.floor(random() * values.length)];
}

function randomInt(random, min, max) {
  return Math.floor(random() * (max - min + 1)) + min;
}

function randomFloat(random, min, max, digits = 2) {
  return Number((min + random() * (max - min)).toFixed(digits));
}

function flattenParts(parts = []) {
  const flattened = [];
  for (const part of parts) {
    if (!part || typeof part !== 'object') continue;
    flattened.push(part);
    for (const nested of Array.isArray(part.strval) ? part.strval : []) {
      if (nested && typeof nested === 'object' && Array.isArray(nested.parts)) {
        flattened.push(...flattenParts(nested.parts));
      }
    }
  }
  return flattened;
}

function cleanSkillName(skillName) {
  return String(skillName ?? '')
    .replace('[単独発動]', '')
    .split('[')[0]
    .split('(')[0]
    .split('（')[0]
    .trim();
}

function calculateFixtureExpected(input, data, spMapping) {
  const skillName = input.skill.name;
  const cleanName = cleanSkillName(skillName);
  let realSkill = data.skills.find((skill) => skill.name === cleanName) ?? null;
  if (!realSkill && skillName.includes('通常攻撃')) {
    realSkill = data.skills.find((skill) => skill.name === '通常攻撃' && String(skill.id).endsWith('01')) ?? null;
  }
  if (!realSkill && skillName.includes('追撃')) {
    realSkill = data.skills.find((skill) => skill.name === '追撃' && String(skill.id).endsWith('91')) ?? null;
  }

  const mappingInfo = spMapping[skillName] ?? spMapping[cleanName] ?? null;
  let sp = 4;
  if (mappingInfo?.sp !== undefined && mappingInfo.sp !== null && mappingInfo.sp !== '-') {
    sp = Number.parseFloat(mappingInfo.sp);
  } else if (realSkill) {
    sp = Number.parseFloat(realSkill.sp_cost ?? realSkill.spCost ?? 4);
  }

  let isNormalAttack = skillName.includes('通常攻撃');
  let isPursuit = skillName.includes('追撃');
  if (mappingInfo) {
    if (mappingInfo.is_normal_attack !== undefined) {
      isNormalAttack = Boolean(mappingInfo.is_normal_attack);
    }
    if (mappingInfo.is_pursuit !== undefined) {
      isPursuit = Boolean(mappingInfo.is_pursuit);
    }
  }

  const mockSkill = realSkill
    ? {
        ...realSkill,
        sp_cost: isNormalAttack || isPursuit ? 0 : sp,
      }
    : {
        id: isNormalAttack ? 46001101 : (isPursuit ? 46001191 : 99999999),
        name: cleanName,
        sp_cost: isNormalAttack || isPursuit ? 0 : sp,
        hit_count: 1,
        parts: [{ skill_type: 'AttackSkill', multipliers: { dr: 13.8 } }],
      };

  const normalizedInput = {
    ...input,
    attacker: {
      ...input.attacker,
      accessoryDestructionRateBonus: Number(input.attacker?.accessoryDestructionRateBonus ?? 0),
    },
    autoBreak: true,
    skill: {
      ...input.skill,
      skillId: mockSkill.id,
      name: mockSkill.name,
      spCostOverride: sp,
      isNormalAttack,
      isPursuit,
    },
  };

  return calculateDestruction(normalizedInput, {
    styles: data.styles,
    enemies: data.enemies,
    skills: [mockSkill],
  });
}

function buildCases() {
  const random = createPrng(SEED);
  const data = loadDamageCalculationData(ROOT_DIR);
  const spMapping = JSON.parse(readFileSync(SKILL_MAPPING_PATH, 'utf8'));
  const styles = [...data.styles].sort((left, right) => Number(left.id) - Number(right.id));
  const enemies = [...data.enemies].sort((left, right) => Number(left.id) - Number(right.id));
  const attackSkills = data.skills
    .filter((skill) => flattenParts(skill.parts).some((part) => ATTACK_SKILL_TYPES.has(part.skill_type)))
    .sort((left, right) => Number(left.id) - Number(right.id));
  const skills = [
    ...attackSkills,
    { id: 46001101, name: '通常攻撃' },
    { id: 46001191, name: '追撃' },
  ];

  if (styles.length === 0 || enemies.length === 0 || attackSkills.length === 0) {
    throw new Error('Required styles, enemies, or attack skills are empty.');
  }

  return Array.from({ length: CASE_COUNT }, (_, index) => {
    const style = pick(random, styles);
    const enemy = pick(random, enemies);
    const skill = pick(random, skills);
    const statusEffects = [];
    const buffCount = randomInt(random, 0, 3);
    for (let buffIndex = 0; buffIndex < buffCount; buffIndex++) {
      statusEffects.push({
        statusType: 'DestructionUp',
        power: randomFloat(random, 5, 120),
        skillName: `BuffSkill${randomInt(random, 1, 5)}`,
      });
    }
    if (random() < 0.3) {
      statusEffects.push({
        statusType: pick(random, ['AttackUp', 'CritRateUp', 'MindEye']),
        power: 30,
        skillName: 'IgnoredSkill',
      });
    }
    const defenderStatusEffects = random() < 0.3
      ? [{ statusType: 'DefenseDown', power: 30, skillName: 'DebuffSkill' }]
      : [];
    const hits = Array.from({ length: randomInt(random, 1, 12) }, () => ({
      damage: randomFloat(random, 0, 150000),
      isMultiHit: random() < 0.5,
      hitRatio: randomFloat(random, 0.05, 1),
    }));
    const input = {
      attacker: {
        characterId: style.chara ?? style.chara_label ?? 'Unknown',
        styleId: style.id,
        accessories: [],
        accessoryDestructionRateBonus: random() < 0.2 ? pick(random, [0.10, 0.12, 0.15]) : 0,
        statusEffects,
      },
      defender: {
        enemyId: enemy.id,
        enemyName: enemy.name ?? 'Unknown',
        destructionRate: randomFloat(random, 1, 3, 4),
        destructionLimit: random() < 0.5 ? null : randomFloat(random, 3, 5),
        dp: randomFloat(random, 0, 300000),
        destructionResist: randomFloat(random, 0, 0.5, 4),
        statusEffects: defenderStatusEffects,
      },
      skill: {
        skillId: skill.id,
        name: skill.name,
      },
      hits,
      autoBreak: true,
    };

    return {
      name: `Deterministic Case ${String(index).padStart(4, '0')}`,
      input,
      expected: calculateFixtureExpected(input, data, spMapping),
    };
  });
}

function buildMetadata(fixtureText) {
  const syncMetadata = JSON.parse(readFileSync(SYNC_METADATA_PATH, 'utf8'));
  return {
    schemaVersion: 1,
    generatorVersion: GENERATOR_VERSION,
    generator: 'scripts/generate-destruction-large-fixtures.mjs',
    generationCommand: 'npm run generate:calc-fixtures',
    checkCommand: 'npm run check:calc-fixtures',
    algorithm: 'mulberry32',
    seed: SEED,
    caseCount: CASE_COUNT,
    inputDataVersion: {
      generatedAt: syncMetadata.generated_at,
      datasets: Object.fromEntries(
        ['styles.json', 'enemies.json', 'skills.json'].map((name) => [
          name,
          syncMetadata.datasets?.[name]?.sha256 ?? null,
        ])
      ),
      skillSpMappingSha256: hashFile('tests/fixtures/skill_sp_mapping.json'),
    },
    calcCoreSha256: Object.fromEntries(
      CALC_CORE_FILES.map((file) => [file, hashFile(file)])
    ),
    fixtureSha256: sha256(fixtureText),
  };
}

function generateArtifacts() {
  const cases = buildCases();
  if (cases.length !== CASE_COUNT) {
    throw new Error(`Expected ${CASE_COUNT} cases, generated ${cases.length}.`);
  }
  const fixtureText = JSON.stringify(cases);
  const metadataText = JSON.stringify(buildMetadata(fixtureText));
  return { fixtureText, metadataText };
}

function checkFile(filePath, expectedText) {
  if (!existsSync(filePath)) return false;
  return readFileSync(filePath, 'utf8') === expectedText;
}

const writeMode = process.argv.includes('--write');
const checkMode = process.argv.includes('--check');
if (writeMode === checkMode) {
  throw new Error('Specify exactly one of --write or --check.');
}

const { fixtureText, metadataText } = generateArtifacts();
if (writeMode) {
  writeFileSync(FIXTURE_PATH, fixtureText);
  writeFileSync(METADATA_PATH, metadataText);
  console.log(`Generated ${CASE_COUNT} deterministic destruction fixtures (seed=${SEED}).`);
} else {
  const fixtureMatches = checkFile(FIXTURE_PATH, fixtureText);
  const metadataMatches = checkFile(METADATA_PATH, metadataText);
  if (!fixtureMatches || !metadataMatches) {
    console.error('Deterministic destruction fixtures are stale. Run: npm run generate:calc-fixtures');
    process.exitCode = 1;
  } else {
    console.log(`Verified ${CASE_COUNT} deterministic destruction fixtures (seed=${SEED}).`);
  }
}
